qr-code/src/main/java/com/adyen/checkout/qrcode/FullQRCodeView.kt

```diff
@@ -0,0 +1,194 @@
+            .onEach { handleEventFlow(it) }
+            .launchIn(coroutineScope)
+    }
+
+    private fun outputDataChanged(outputData: QRCodeOutputData) {
+        Logger.d(TAG, "outputDataChanged")
+
+        updateMessageText(outputData.paymentMethodType)
+        updateLogo(outputData.paymentMethodType)
+        updateQrImage(outputData.qrImageUrl)
+        updateAmount(delegate.componentParams)
+    }
+
+    private fun updateAmount(componentParams: ComponentParams) {
+        if (!componentParams.amount.isEmpty) {
+            val formattedAmount = CurrencyUtils.formatAmount(
+                componentParams.amount,
+                componentParams.shopperLocale
+            )
+            binding.textviewAmount.isVisible = true
+            binding.textviewAmount.text = formattedAmount
+        } else {
+            binding.textviewAmount.isVisible = false
+        }
+    }
+
+    private fun updateMessageText(paymentMethodType: String?) {
+        val resId = getMessageTextResource(paymentMethodType) ?: return
+        binding.textViewTopLabel.text = localizedContext.getString(resId)
+    }
+
+    private fun updateLogo(paymentMethodType: String?) {
+        if (!paymentMethodType.isNullOrEmpty()) {
+            binding.imageViewLogo.loadLogo(
+                environment = delegate.componentParams.environment,
+                txVariant = paymentMethodType,
+                size = LogoSize.LARGE
+            )
+        }
+    }
+
+    private fun updateQrImage(qrImageUrl: String?) {
+        if (!qrImageUrl.isNullOrEmpty()) {
+            binding.imageViewQrcode.load(url = qrImageUrl)
+        }
+    }
+
+    @StringRes
+    private fun getMessageTextResource(paymentMethodType: String?): Int? {
+        return when (paymentMethodType) {
+            PaymentMethodTypes.PAY_NOW -> R.string.checkout_qr_code_pay_now
+            else -> null
+        }
+    }
+
+    private fun onTimerTick(timerData: TimerData) {
+        val minutes = TimeUnit.MILLISECONDS.toMinutes(timerData.millisUntilFinished)
+        val seconds = TimeUnit.MILLISECONDS.toSeconds(timerData.millisUntilFinished) % TimeUnit.MINUTES.toSeconds(1)
+
+        val minutesSecondsString = localizedContext.getString(
+            R.string.checkout_qr_code_time_left_format,
+            minutes,
+            seconds
+        )
+
+        binding.textViewTimer.text = localizedContext.getString(
+            R.string.checkout_qr_code_pay_now_timer_text,
+            minutesSecondsString
+        )
+        binding.progressIndicator.progress = timerData.progress
+    }
+
+    private fun handleEventFlow(event: QrCodeUIEvent) {
+        when (event) {
+            QrCodeUIEvent.QrImageDownloadedResult.Success -> {
+                // TODO: add translation
+                context.toast("qr code downloaded successfully")
+            }
+            is QrCodeUIEvent.QrImageDownloadedResult.Failure -> {
+                Logger.d(TAG, "download file failed", event.throwable)
```

Comment:
Did we consider showing anything to the user when the download fails?

Reply:
Good catch, will add a message for it.